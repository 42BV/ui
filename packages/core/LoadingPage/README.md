# @42.nl/ui-core-loading-page

## Installation

Install the package in your project directory with:

```sh
// with npm
npm install @42.nl/ui-core-loading-page

// with yarn
yarn add @42.nl/ui-core-loading-page
```
