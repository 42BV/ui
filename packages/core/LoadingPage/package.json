{
  "name": "@42.nl/ui-core-loading-page",
  "version": "0.1.0-alpha.0",
  "description": "LoadingPage React component as used in 42.nl products.",
  "main": "lib/LoadingPage.js",
  "scripts": {
    "dev": "STORYBOOK_MODE=isolated STORYBOOK_COMPONENT=$(pwd) start-storybook -c ../../../docs -p 9000",
    "test": "jest --config=jest.config.js",
    "build": "yarn clean && tsc && yarn copy:scss",
    "copy:scss": "rsync -a src/*.scss lib/scss/",
    "lint": "eslint src --ext js,jsx,ts,tsx --max-warnings=0",
    "clean": "rm -rf lib"
  },
  "files": [
    "lib/*"
  ],
  "dependencies": {
    "@42.nl/ui-core-spinner": "^0.1.0-alpha.0",
    "classnames": "2.2.6"
  },
  "peerDependencies": {
    "react": "^16.8.0",
    "reactstrap": "^8.0.1"
  },
  "devDependencies": {
    "react": "16.8.6",
    "reactstrap": "8.0.1"
  },
  "keywords": [
    "ui",
    "react",
    "bootstrap",
    "infobadge"
  ],
  "contributors": [
    "Maarten Hus <maarten.hus@42.nl>",
    "Pouja Nikray <pouja.nikray@42.nl>",
    "Jeffrey van Hoven <jeffrey.van.hoven@42.nl> (https://jvhoven.nl)"
  ],
  "homepage": "https://42bv.github.io/ui",
  "bugs": {
    "url": "https://github.com/42BV/ui/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/42BV/ui.git",
    "directory": "packages/core/InfoBadge"
  },
  "license": "Apache-2.0",
  "publishConfig": {
    "access": "public"
  }
}
