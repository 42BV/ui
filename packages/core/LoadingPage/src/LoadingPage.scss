// @remove-on-build-start
@import '~@42.nl/ui-core-styling/lib/variables';
@import '~@42.nl/ui-core-styling/lib/bootstrap';
@import '~@42.nl/ui-core-spinner/lib/scss/Spinner';
// @remove-on-build-end

.loading-page {
  margin: 0;
  padding: 0;
  background-color: $body-bg;
}
