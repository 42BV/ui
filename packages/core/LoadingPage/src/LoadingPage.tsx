import React, { useEffect, useState } from 'react';
import classNames from 'classnames';
import Spinner from '@42.nl/ui-core-spinner';

interface Props {
  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
}

/**
 * The LoadingPage component embeds the Spinner component in a page-friendly matter.
 *
 * Use this for showing a loading indicator when navigating to pages that fetch data.
 */
export default function LoadingPage({ className }: Props) {
  const size = 150;

  const [showSpinner, setShowSpinner] = useState(false);
  useEffect(() => {
    const timeoutId = window.setTimeout(() => {
      setShowSpinner(true);
    }, 200);

    return () => {
      window.clearTimeout(timeoutId);
    };
  }, []);

  return (
    <div className={classNames('LoadingPage', className)}>
      <div
        className="vh-100 d-flex flex-column justify-content-center align-items-center"
        style={{ marginTop: -(size / 2) }}
      >
        {showSpinner && <Spinner size={size} color="#f0ad4e" />}
      </div>
    </div>
  );
}
