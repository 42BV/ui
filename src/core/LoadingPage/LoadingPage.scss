.loading-page {
  margin: 0;
  padding: 0;
  background-color: var(--bs-body-bg);
}
