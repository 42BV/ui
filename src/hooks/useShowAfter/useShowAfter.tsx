import { useEffect, useState } from 'react';

/**
 * Returns `false` but after a predetermined `after` time returns
 * `true`.
 *
 * Useful for wanting to show content only after a time. For example
 * to not show a loading spinner to soon, only
 * after it takes a while.
 *
 * @param after The amount of time in milliseconds after which the return value should become `true`.
 */
export function useShowAfter(after: number): boolean {
  const [show, setShow] = useState(false);
  useEffect(() => {
    const timeoutId = window.setTimeout(() => {
      setShow(true);
    }, after);

    return () => {
      window.clearTimeout(timeoutId);
    };
  }, [after]);

  return show;
}
