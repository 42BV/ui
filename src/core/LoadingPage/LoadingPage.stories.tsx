import React from 'react';

import { LoadingPage } from './LoadingPage';

export default {
  title: 'core/LoadingPage',

  parameters: {
    component: LoadingPage
  }
};

const DefaultStory = () => {
  return <LoadingPage className="mt-0" />;
};

export const Default = {
  render: DefaultStory,
  name: 'default'
};

const CustomHeightStory = () => {
  return <LoadingPage height={100} />;
};

export const CustomHeight = {
  render: CustomHeightStory,
  name: 'custom height'
};

const CustomStyleStory = () => {
  return <LoadingPage style={{ backgroundColor: 'red' }} />;
};

export const CustomStyle = {
  render: CustomStyleStory,
  name: 'custom style'
};
