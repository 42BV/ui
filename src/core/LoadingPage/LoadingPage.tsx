import classNames from 'classnames';
import React from 'react';
import { Spinner } from '../Spinner/Spinner';
import { useShowAfter } from '../../hooks/useShowAfter/useShowAfter';
import { UIBasePropsWithCSSProperties } from '../types';

type Props = {
  /**
   * Optionally a height, by default this will be the full height
   * of the view port.
   */
  height?: number;
} & Partial<UIBasePropsWithCSSProperties>;

/**
 * The LoadingPage component embeds the Spinner component in a page-friendly matter.
 *
 * Use this for showing a loading indicator when navigating to pages that fetch data.
 */
export function LoadingPage({ className, style, height }: Props) {
  const size = 150;

  const showSpinner = useShowAfter(200);

  return (
    <div className={classNames('loading-page', className)} style={style}>
      <div
        className={classNames(
          'd-flex flex-column justify-content-center align-items-center',
          { 'vh-100': height === undefined }
        )}
        style={{
          marginTop: height === undefined ? -(size / 2) : undefined,
          height
        }}
      >
        {showSpinner && <Spinner size={size} color="#f0ad4e" />}
      </div>
    </div>
  );
}
