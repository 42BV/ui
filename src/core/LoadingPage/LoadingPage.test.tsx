import React from 'react';
import { vi } from 'vitest';
import { render, screen } from '@testing-library/react';

import * as ShowAfter from '../../hooks/useShowAfter/useShowAfter';

import { LoadingPage } from './LoadingPage';

describe('Component: LoadingPage', () => {
  function setup({ show, height }: { show: boolean; height?: number }) {
    vi.spyOn(ShowAfter, 'useShowAfter').mockReturnValue(show);

    const { container } = render(<LoadingPage height={height} />);

    return { container };
  }

  describe('ui', () => {
    test('default', () => {
      const { container } = setup({ show: true });

      expect(container).toMatchSnapshot();
    });

    test('with no spinner because it is not after the timeout', () => {
      setup({ show: false });

      expect(screen.queryByRole('graphics-document')).toBeNull();
    });

    test('with custom height', () => {
      const { container } = setup({ show: true, height: 200 });

      expect(container).toMatchSnapshot();
    });
  });
});
