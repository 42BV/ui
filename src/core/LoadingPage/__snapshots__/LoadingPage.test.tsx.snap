// Vitest Snapshot v1

exports[`Component: LoadingPage > ui > default 1`] = `
<div>
  <div
    class="loading-page"
  >
    <div
      class="d-flex flex-column justify-content-center align-items-center vh-100"
      style="margin-top: -75px;"
    >
      <svg
        class="spinner"
        height="150"
        viewBox="25 25 50 50"
        width="150"
      >
        <circle
          class="path"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          stroke="#f0ad4e"
          stroke-miterlimit="10"
          stroke-width="4"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Component: LoadingPage > ui > with custom height 1`] = `
<div>
  <div
    class="loading-page"
  >
    <div
      class="d-flex flex-column justify-content-center align-items-center"
      style="height: 200px;"
    >
      <svg
        class="spinner"
        height="150"
        viewBox="25 25 50 50"
        width="150"
      >
        <circle
          class="path"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          stroke="#f0ad4e"
          stroke-miterlimit="10"
          stroke-width="4"
        />
      </svg>
    </div>
  </div>
</div>
`;
